posts/python-decorators.md

```diff
@@ -0,0 +1,180 @@
+Decorators are a fairly common thing in python, moreso when using frameworks such as flask or click, but many only know how to them, and not how to write their own.
+
+# How it works
+When you define a function in python, that function becomes an object. 
+Let's say you write a function called `hello_world` in python. That function is actually just an object called `hello_world` that is a callable. This is very similar to how lambdas work: `hello = lambda: print('hello')` would be the same as `def hello(): print('hello')`. Python decorators work by wrapping an object at definition time, and returning a new object definition. Essentially, if you were decorating a lambda, it would be the equivalent of `hello = decorate(lambda: print('hello'))`. Decorate is passed the function -- which it can use however it wants -- then returns another object. The sky is the limit. The decorator has the ability to swallow the function, or return something that is not a function, if it wanted. 
+
+Here is an example of a decorator that does useless things. It simply swallows the function it is given entirely, and always returns 5. This is mostly never what you want, but it's a good example of a decorator that is simply passed a function, and returns an object. 
+
+```python
+@my_decorator_that_just_returns_5
+def hello():
+    print('hello')
+
+>>> hello()
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+TypeError: 'int' object is not callable
+'int' object is not callable
+
+>>> hello
+5
+```
+
+
+# Writing your own decorator
+As mentioned, a decorator is simply a function that is passed a function, and returns an object. So, to start writing a decorator, we just need to define a function. In most cases, we want the object our decorator returns, to actually mimic the function we decorated. This means that the object the decorator returns, needs to be a function itself. For example, let's say we wanted to simply print every time the function was called, we could write a function that prints that information, then calls the function. But that function would need to be returned by the decorator. This usually leads to function nesting such as:
+
+```python
+def mydecorator(f):  # f is the function passed to us from python
+    def log_f_as_called():
+        print(f'{f} was called.')
+        f()
+    return log_f_as_called
+```
+
+As you can see, we have a nested function definition, and the decorator function is returning that function. This way, the function `hello` can still be called like a standard function, the caller does not need to know that it is decorated. We can now define `hello` as the following:
+
+```python
+@mydecorator
+def hello():
+    print('hello')
+```
+
+We get the following output:
+```python
+>>> hello()
+called <function hello at 0x7f27738d7510>
```

Comment:
When I run the code, I get this output: `<function hello at 0x10f9240d0> was called.`

Reply:
I don't think this was fixed.